README.md
# qsfs

[![Build Status](https://travis-ci.org/yunify/qsfs-fuse.svg?branch=master)][build link]
[![License](http://img.shields.io/badge/license-apache%20v2-blue.svg)][license link]

**qsfs** is a FUSE based filesystem that allows you to mount a qingstor bucket in Linux.


## Features

- Large subset of POSIX including reading/writing files, directories, symlinks, .etc,
  (chmod/chown/utimens will be supported later).
- File renames via server-side move.
- File Transfer:
  - Large files uploads via multipart parallel uploads.
  - Large files downloads via parallel byte-range downloads.
  - Large files transfer in chunks (10 MB chunks by default). If you are uploading large files (e.g. larger than 1 GB), you can increase the transfer buffer size and the max
  parallel transfers.
- Data integrity via MD5 hashes.
- Cache:
  - In-memory metadata caching.
  - In-memory file data caching. For a big file, partial file data may been stored in a local disk file when the im-memory file cache is not available.
- Logging/Debugging:
  - Support to log messages to console or to a directory.
  - Support turn on debug message to log by specifying debug option *-d*, this option
  will also enable FUSE debug mode.
  - You can turn on debug message from curl by specifying option *-U*.


## Installation

See the [INSTALL][install link] for installation instructions.


## Usage

Enter your QingCloud API access keys in a file `/path/to/cred`, go to [QingCloud Console][qingcloud console link] to create them if you do not have one:
```sh
 $ echo YourAccessKeyId:YourSecretKey > /path/to/cred
```

Make sure the file `/path/to/cred` has proper permissions (if you get 'permissions' error when mounting):
```sh
 $ chmod 600 /path/to/cred
```

Run qsfs with an existing bucket `mybucket` and directory `/path/to/mountpoint`:
```sh
 $ [sudo] qsfs mybucket /path/to/mountpoint -c=/path/to/cred
```

If you encounter any errors, enable debug output:
```sh
 $ [sudo] qsfs mybucket /path/to/mountpoint -c=/path/to/cred -d
```

You can also mount on boot by entering the following line to /etc/fstab:

```sh
qsfs#mybucket /path/to/mountpoint fuse _netdev,-z=pek3a,-c=/path/to/cred,allow_other 0 0
```

> Notice: in /etc/fstab, you need to use short options (for example, use short option `-c` instead of long option `--credentials`) in order to pass them to qsfs.

You can log messages to console:
```sh
 $ [sudo] qsfs mybucket /path/to/mountpoint -c=/path/to/cred -f
```

Or you can log messages to log file by specifying a directory `/path/to/logdir/`:
```sh
 $ [sudo] qsfs mybucket /path/to/mountpoint -c=/path/to/cred -l=/path/to/logdir/
```

Specify log level (INFO,WARN,ERROR and FATAL):
```sh
 $ [sudo] qsfs mybucket /path/to/mountpoint -c=/path/to/cred -L=INFO -d
```

To umount:
```sh
 $ [sudo] fusermount -uqz /path/to/mountpoint
```
  or
```sh
 $ [sudo] umount -l /path/to/mountpoint
```

For help:
```sh
 $ qsfs -h
```

## Command line options
Supported general options are listed as following,

| short | full | type | required | usage |
| ----- |------|:------:|:----------:|------ |
| -z | --zone        | string  | Y | Specify zone or region
| -c | --credentials | string  | N | Specify credentials file, default path is `/etc/qsfs.cred`
| -l | --logdir      | string  | N | Specify log directory, default path is `/var/log/qsfs/`
| -L | --loglevel    | string  | N | Specify min log level (INFO,WARN,ERROR or FATAL), message lower than this level don't logged; default value is `INFO`.
| -F | --filemode    | octal   | N | Specify the permission bits in st_mode for file objects without x-qs-meta-mode header. The value is given in octal representation; default value is `0644`
| -D | --dirmode     | octal   | N | Specify the permission bits in st_mode for directory objects without x-qs-meta-mode header. The value is given in octal representation; default value is `0755`
| -u | --umaskmp     | octal   | N | Specify the permission bits in st_mode for the mount point directory. This option only works when you set with the fuse allow_other option. The resulting permission bits are the ones missing from the given umask value. The value is given in octal representation; default value is `0000`
| -r | --retries     | integer | N | Specify number of times to retry a failed transaction, default value is `3 times`
| -R | --reqtimeout  | integer | N | Specify time (seconds) to wait before giving up connection, default value is `300 seconds`
| -Z | --maxcache    | integer | N | Specify max in-memory cache size (MB) for files, default value is `200 MB`
| -k | --diskdir     | string  | N | Specify the directory to store file data when in-memory cache is not availabe, default path is `/tmp/qsfs_cache/`
| -t | --maxstat     | integer | N | Specify max count (K) of cached stat entrys, default value is `2 K`
| -e | --statexpire  | integer | N | Specify expire time (minutes) for stat entries, negative value will disable stat expire, default is no expire
| -i | --maxlist     | integer | N | Specify max count of files of ls operation. A value of zero will list all files, default is to list all files
| -y | --fscap       | integer | N | Specify filesystem capacity (GB), default value is `1PB`
| -n | --numtransfer | integer | N | Specify max number file tranfers to run in parallel, you can increase the value when transfer large files, default value is `5`
| -b | --bufsize     | integer | N | Specify file transfer buffer size (MB), this should be larger than 8MB, default value is `10 MB`
| -j | --prefetchsize| integer | N | Specify file read max prefetch size (MB), default value is `20 MB`
| -H | --host        | string  | N | Specify host name, default value is `qingstor.com`
| -p | --protocol    | string  | N | Specify protocol (https or http) default value is `https`
| -P | --port        | integer | N | Specify port, default is 443 for https and 80 for http

Supported miscellaneous options are list as following,

| short | full | type | required | usage |
| ----- |------|:------:|:----------:|------ |
| -J | --prefetch    | bool | N | Enable read prefetch
| -m | --contentMD5  | bool | N | Enable writes with MD5 hashs to ensure data integrity
| -K | --keeplogdir  | bool | N | Do not clear log directory at beginning
| -f | --forground   | bool | N | Turn on log to STDERR and enable FUSE foreground mode
| -s | --single      | bool | N | Turn on FUSE single threaded option - disable multi-threaded
| -d | --debug       | bool | N | Turn on debug messages to log
| -g | --fusedbg     | bool | N | Enable FUSE debug option
| -U | --curldbg     | bool | N | Turn on debug message from curl
| -h | --help        | bool | N | Print qsfs help
| -V | --version     | bool | N | Print qsfs version

You can also specify FUSE specific mount options with `-o opt [,opt...]` e.g. nonempty, allow_other, etc. See the FUSE's README for the full set.


## Limitations

Generally qingstor cannot offer the same performance or semantics as a local file system.  More specifically:
- Random writes or appends to files require rewriting the entire file
- Metadata operations such as listing directories have poor performance due to network latency
- [Eventual consistency][eventual consistency wiki] can temporarily yield stale data
- No atomic renames of files or directories
- No coordination between multiple clients mounting the same bucket
- No hard links


## Frequently Asked Questions

- [FAQ wiki page][faq wiki link]


## Support

If you notice any issue, please open an [issue][issue link] on GitHub. Please search the existing issues and see if others are also experiencing the issue before opening a new issue. Please include the version of qsfs, Compiler version, CMake version, and OS you’re using.


## License

See the [LICENSE][license link] for details. In summary, qsfs is licensed under the Apache License (Version 2.0, January 2004).


[build link]: https://travis-ci.org/yunify/qsfs-fuse
[eventual consistency wiki]: https://en.wikipedia.org/wiki/Eventual_consistency
[faq wiki link]: https://github.com/yunify/qsfs-fuse/wiki/FAQ
[install link]: https://github.com/yunify/qsfs-fuse/blob/master/INSTALL.md
[issue link]: https://github.com/yunify/qsfs-fuse/issues
[license link]: https://github.com/yunify/qsfs-fuse/blob/master/COPYING
[qingcloud console link]: https://console.qingcloud.com/access_keys/

ChangeLog
* Mon Dec 18 2018 - Yunify Inc.

- qsfs: version 1.0.10
- Add·backward·compatiblility for ListObjects.

* Thu Dec 6 2018 - Yunify Inc.

- qsfs: version 1.0.9
- Update ListObjects.

* Fri Oct 19 2018 - Yunify Inc.

- qsfs: version 1.0.8
- Fix Cache::Free to avoid list::erase invalid iterator.
- Add option max log size.
- Other bugfix.

* Mon Sep 3 2018 - Yunify Inc.
- qsfs: version 1.0.7
- Fix Cache::Free list::erase segment fault.

* Tue Jul 24 2018 - Yunify Inc.
- qsfs: version 1.0.6
- Fix memory leak when download file.
- Change log default path to /var/log/qsfs.

* Tue Jun 12 2018 - Yunify Inc.

- qsfs: version 1.0.5
- Add option to specify file system size (capacity)
- Add null implementation for meta related api (such as chmod, chown and utimens)
- Set user agent to fixed value of qsfs-version.number
- Add read prefetch option
- Ajust log message

* Sat May 26 2018 - Yunify Inc.

- qsfs: version 1.0.4
- Refactor to ensure file ops (read, write, truncate, .etc) thread safe
- Partially remove Clinet dependency on DirectoryTree and Cache
- Add individual folder for each integration test to decrease the coupling among them
- Replace static dynamic container (set, map) with array
- Fix issue of failure to pass host to sdk

* Fri Apr 13 2018 - Yunify Inc.

- qsfs: version 1.0.3
- Refactor Directory Tree Invalidation/Remove
- Fix Directory Tree Rename issue which not rename directory descendants
- Add integration test framework
- Add 24 function test cases including write, read, rename, seek, truncate, rsync, redirect, symlink, etc.
- Add glog signal failure handler to log failure dump and customize failer writer to convert failure dump address info to readable c++ name.
- Add function name in log, and glog env such as max_log_size, clorlogtostderr, .etc
- Seperate debug option for qsfs and fuse 
- Fix failure of function test 'rename file before close'
- Log to stderr if umaskmp is not set with allow_other together
- Not to set content type in QSClient as QingStor server will handle file content type
- Link to static QingStor SDK library
- Refactor to remove mtime from cache
- Add File info to log message

- qsfs: version 1.0.2
- To synchronize Cache operations to avoid data race
- Fix failure to truncate a new created empty file
- Fix failure to upload big file with odd bytes
- Fix wrong data position when read from cache
- No log message in Drive Cleanup to avoid dead reference to Log
- Set connection timeout to 300s as same as curl default
- Refactor to gurantee data consistency
- Add more log messages
- Add CMake option UPDATE_CONNECT to automatically update external project from remote repo

* Sun Mar 25 2018 - Yunify Inc.

- qsfs: version 1.0.1
- Support to enable writes with MD5 hashs to ensure data integrity
- Add option to specify the permission bits in st_mode for file/directory objects without x-qs-meta-mode header
- Add option to umask the permission bits in st_mode for the mount point diretory
- Support fuse options of uid, gid and umask
- Prevent rpmbuild from stripping debuginfo when specify CMAKE_BUILD_TYPE with Debug
- Modify some log message level
- Fix some lint warnings

* Mon Feb 5 2018 - Yunify Inc.

- qsfs: version 1.0.0


src/base/MD5.cpp
/* MD5
 converted to C++ class by Frank Thilo (thilo@unix-ag.org)
 for bzflag (http://www.bzflag.org)

   based on:

   md5.h and md5.c
   reference implemantion of RFC 1321

   Copyright (C) 1991-2, RSA Data Security, Inc. Created 1991. All
rights reserved.

License to copy and use this software is granted provided that it
is identified as the "RSA Data Security, Inc. MD5 Message-Digest
Algorithm" in all material mentioning or referencing this software
or this function.

License is also granted to make and use derivative works provided
that such works are identified as "derived from the RSA Data
Security, Inc. MD5 Message-Digest Algorithm" in all material
mentioning or referencing the derived work.

RSA Data Security, Inc. makes no representations concerning either
the merchantability of this software or the suitability of this
software for any particular purpose. It is provided "as is"
without express or implied warranty of any kind.

These notices must be retained in any copies of any part of this
documentation and/or software.

*/

/* interface header */
#include "MD5.h"

/* system implementation headers */
#include <cstdio>
#include <sstream>

// Constants for MD5Transform routine.
#define S11 7
#define S12 12
#define S13 17
#define S14 22
#define S21 5
#define S22 9
#define S23 14
#define S24 20
#define S31 4
#define S32 11
#define S33 16
#define S34 23
#define S41 6
#define S42 10
#define S43 15
#define S44 21

///////////////////////////////////////////////

// F, G, H and I are basic MD5 functions.
inline MD5::uint4 MD5::F(uint4 x, uint4 y, uint4 z) { return x & y | ~x & z; }

inline MD5::uint4 MD5::G(uint4 x, uint4 y, uint4 z) { return x & z | y & ~z; }

inline MD5::uint4 MD5::H(uint4 x, uint4 y, uint4 z) { return x ^ y ^ z; }

inline MD5::uint4 MD5::I(uint4 x, uint4 y, uint4 z) { return y ^ (x | ~z); }

// rotate_left rotates x left n bits.
inline MD5::uint4 MD5::rotate_left(uint4 x, int n) {
  return (x << n) | (x >> (32 - n));
}

// FF, GG, HH, and II transformations for rounds 1, 2, 3, and 4.
// Rotation is separate from addition to prevent recomputation.
inline void MD5::FF(uint4 &a, uint4 b, uint4 c, uint4 d, uint4 x, uint4 s,
                    uint4 ac) {
  a = rotate_left(a + F(b, c, d) + x + ac, s) + b;
}

inline void MD5::GG(uint4 &a, uint4 b, uint4 c, uint4 d, uint4 x, uint4 s,
                    uint4 ac) {
  a = rotate_left(a + G(b, c, d) + x + ac, s) + b;
}

inline void MD5::HH(uint4 &a, uint4 b, uint4 c, uint4 d, uint4 x, uint4 s,
                    uint4 ac) {
  a = rotate_left(a + H(b, c, d) + x + ac, s) + b;
}

inline void MD5::II(uint4 &a, uint4 b, uint4 c, uint4 d, uint4 x, uint4 s,
                    uint4 ac) {
  a = rotate_left(a + I(b, c, d) + x + ac, s) + b;
}

//////////////////////////////////////////////

// default ctor, just initailize
MD5::MD5() { init(); }

//////////////////////////////////////////////

// nifty shortcut ctor, compute MD5 for string and finalize it right away
MD5::MD5(const std::string &text) {
  init();
  update(text.c_str(), text.length());
  finalize();
}

//////////////////////////////

void MD5::init() {
  finalized = false;

  count[0] = 0;
  count[1] = 0;

  // load magic initialization constants.
  state[0] = 0x67452301;
  state[1] = 0xefcdab89;
  state[2] = 0x98badcfe;
  state[3] = 0x10325476;
}

//////////////////////////////

// decodes input (unsigned char) into output (uint4). Assumes len is a multiple
// of 4.
void MD5::decode(uint4 output[], const uint1 input[], size_type len) {
  for (unsigned int i = 0, j = 0; j < len; i++, j += 4)
    output[i] = ((uint4)input[j]) | (((uint4)input[j + 1]) << 8) |
                (((uint4)input[j + 2]) << 16) | (((uint4)input[j + 3]) << 24);
}

//////////////////////////////

// encodes input (uint4) into output (unsigned char). Assumes len is
// a multiple of 4.
void MD5::encode(uint1 output[], const uint4 input[], size_type len) {
  for (size_type i = 0, j = 0; j < len; i++, j += 4) {
    output[j] = input[i] & 0xff;
    output[j + 1] = (input[i] >> 8) & 0xff;
    output[j + 2] = (input[i] >> 16) & 0xff;
    output[j + 3] = (input[i] >> 24) & 0xff;
  }
}

//////////////////////////////

// apply MD5 algo on a block
void MD5::transform(const uint1 block[blocksize]) {
  uint4 a = state[0], b = state[1], c = state[2], d = state[3], x[16];
  decode(x, block, blocksize);

  /* Round 1 */
  FF(a, b, c, d, x[0], S11, 0xd76aa478);  /* 1 */
  FF(d, a, b, c, x[1], S12, 0xe8c7b756);  /* 2 */
  FF(c, d, a, b, x[2], S13, 0x242070db);  /* 3 */
  FF(b, c, d, a, x[3], S14, 0xc1bdceee);  /* 4 */
  FF(a, b, c, d, x[4], S11, 0xf57c0faf);  /* 5 */
  FF(d, a, b, c, x[5], S12, 0x4787c62a);  /* 6 */
  FF(c, d, a, b, x[6], S13, 0xa8304613);  /* 7 */
  FF(b, c, d, a, x[7], S14, 0xfd469501);  /* 8 */
  FF(a, b, c, d, x[8], S11, 0x698098d8);  /* 9 */
  FF(d, a, b, c, x[9], S12, 0x8b44f7af);  /* 10 */
  FF(c, d, a, b, x[10], S13, 0xffff5bb1); /* 11 */
  FF(b, c, d, a, x[11], S14, 0x895cd7be); /* 12 */
  FF(a, b, c, d, x[12], S11, 0x6b901122); /* 13 */
  FF(d, a, b, c, x[13], S12, 0xfd987193); /* 14 */
  FF(c, d, a, b, x[14], S13, 0xa679438e); /* 15 */
  FF(b, c, d, a, x[15], S14, 0x49b40821); /* 16 */

  /* Round 2 */
  GG(a, b, c, d, x[1], S21, 0xf61e2562);  /* 17 */
  GG(d, a, b, c, x[6], S22, 0xc040b340);  /* 18 */
  GG(c, d, a, b, x[11], S23, 0x265e5a51); /* 19 */
  GG(b, c, d, a, x[0], S24, 0xe9b6c7aa);  /* 20 */
  GG(a, b, c, d, x[5], S21, 0xd62f105d);  /* 21 */
  GG(d, a, b, c, x[10], S22, 0x2441453);  /* 22 */
  GG(c, d, a, b, x[15], S23, 0xd8a1e681); /* 23 */
  GG(b, c, d, a, x[4], S24, 0xe7d3fbc8);  /* 24 */
  GG(a, b, c, d, x[9], S21, 0x21e1cde6);  /* 25 */
  GG(d, a, b, c, x[14], S22, 0xc33707d6); /* 26 */
  GG(c, d, a, b, x[3], S23, 0xf4d50d87);  /* 27 */
  GG(b, c, d, a, x[8], S24, 0x455a14ed);  /* 28 */
  GG(a, b, c, d, x[13], S21, 0xa9e3e905); /* 29 */
  GG(d, a, b, c, x[2], S22, 0xfcefa3f8);  /* 30 */
  GG(c, d, a, b, x[7], S23, 0x676f02d9);  /* 31 */
  GG(b, c, d, a, x[12], S24, 0x8d2a4c8a); /* 32 */

  /* Round 3 */
  HH(a, b, c, d, x[5], S31, 0xfffa3942);  /* 33 */
  HH(d, a, b, c, x[8], S32, 0x8771f681);  /* 34 */
  HH(c, d, a, b, x[11], S33, 0x6d9d6122); /* 35 */
  HH(b, c, d, a, x[14], S34, 0xfde5380c); /* 36 */
  HH(a, b, c, d, x[1], S31, 0xa4beea44);  /* 37 */
  HH(d, a, b, c, x[4], S32, 0x4bdecfa9);  /* 38 */
  HH(c, d, a, b, x[7], S33, 0xf6bb4b60);  /* 39 */
  HH(b, c, d, a, x[10], S34, 0xbebfbc70); /* 40 */
  HH(a, b, c, d, x[13], S31, 0x289b7ec6); /* 41 */
  HH(d, a, b, c, x[0], S32, 0xeaa127fa);  /* 42 */
  HH(c, d, a, b, x[3], S33, 0xd4ef3085);  /* 43 */
  HH(b, c, d, a, x[6], S34, 0x4881d05);   /* 44 */
  HH(a, b, c, d, x[9], S31, 0xd9d4d039);  /* 45 */
  HH(d, a, b, c, x[12], S32, 0xe6db99e5); /* 46 */
  HH(c, d, a, b, x[15], S33, 0x1fa27cf8); /* 47 */
  HH(b, c, d, a, x[2], S34, 0xc4ac5665);  /* 48 */

  /* Round 4 */
  II(a, b, c, d, x[0], S41, 0xf4292244);  /* 49 */
  II(d, a, b, c, x[7], S42, 0x432aff97);  /* 50 */
  II(c, d, a, b, x[14], S43, 0xab9423a7); /* 51 */
  II(b, c, d, a, x[5], S44, 0xfc93a039);  /* 52 */
  II(a, b, c, d, x[12], S41, 0x655b59c3); /* 53 */
  II(d, a, b, c, x[3], S42, 0x8f0ccc92);  /* 54 */
  II(c, d, a, b, x[10], S43, 0xffeff47d); /* 55 */
  II(b, c, d, a, x[1], S44, 0x85845dd1);  /* 56 */
  II(a, b, c, d, x[8], S41, 0x6fa87e4f);  /* 57 */
  II(d, a, b, c, x[15], S42, 0xfe2ce6e0); /* 58 */
  II(c, d, a, b, x[6], S43, 0xa3014314);  /* 59 */
  II(b, c, d, a, x[13], S44, 0x4e0811a1); /* 60 */
  II(a, b, c, d, x[4], S41, 0xf7537e82);  /* 61 */
  II(d, a, b, c, x[11], S42, 0xbd3af235); /* 62 */
  II(c, d, a, b, x[2], S43, 0x2ad7d2bb);  /* 63 */
  II(b, c, d, a, x[9], S44, 0xeb86d391);  /* 64 */

  state[0] += a;
  state[1] += b;
  state[2] += c;
  state[3] += d;

  // Zeroize sensitive information.
  memset(x, 0, sizeof x);
}

//////////////////////////////

// MD5 block update operation. Continues an MD5 message-digest
// operation, processing another message block
void MD5::update(const unsigned char input[], size_type length) {
  // compute number of bytes mod 64
  size_type index = count[0] / 8 % blocksize;

  // Update number of bits
  if ((count[0] += (length << 3)) < (length << 3)) count[1]++;
  count[1] += (length >> 29);

  // number of bytes we need to fill in buffer
  size_type firstpart = 64 - index;

  size_type i;

  // transform as many times as possible.
  if (length >= firstpart) {
    // fill buffer first, transform
    memcpy(&buffer[index], input, firstpart);
    transform(buffer);

    // transform chunks of blocksize (64 bytes)
    for (i = firstpart; i + blocksize <= length; i += blocksize)
      transform(&input[i]);

    index = 0;
  } else
    i = 0;

  // buffer remaining input
  memcpy(&buffer[index], &input[i], length - i);
}

//////////////////////////////

// for convenience provide a verson with signed char
void MD5::update(const char input[], size_type length) {
  update((const unsigned char *)input, length);
}

//////////////////////////////

// MD5 finalization. Ends an MD5 message-digest operation, writing the
// the message digest and zeroizing the context.
MD5 &MD5::finalize() {
  static unsigned char padding[64] = {
      0x80, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
      0,    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
      0,    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0};

  if (!finalized) {
    // Save number of bits
    unsigned char bits[8];
    encode(bits, count, 8);

    // pad out to 56 mod 64.
    size_type index = count[0] / 8 % 64;
    size_type padLen = (index < 56) ? (56 - index) : (120 - index);
    update(padding, padLen);

    // Append length (before padding)
    update(bits, 8);

    // Store state in digest
    encode(digest, state, 16);

    // Zeroize sensitive information.
    memset(buffer, 0, sizeof buffer);
    memset(count, 0, sizeof count);

    finalized = true;
  }

  return *this;
}

//////////////////////////////

// return hex representation of digest as string
std::string MD5::hexdigest() const {
  if (!finalized) return "";

  char buf[33];
  for (int i = 0; i < 16; i++) sprintf(buf + i * 2, "%02x", digest[i]);
  buf[32] = 0;

  return std::string(buf);
}

//////////////////////////////

std::ostream &operator<<(std::ostream &out, MD5 md5) {
  return out << md5.hexdigest();
}

//////////////////////////////

std::string md5(const std::string str) {
  MD5 md5 = MD5(str);

  return md5.hexdigest();
}

std::string md5(const boost::shared_ptr<std::iostream> &stream) {
  std::stringstream ss;
  stream->seekg(0, std::ios_base::beg);
  ss << stream->rdbuf();
  MD5 md5 = MD5(ss.str());
  stream->seekg(0, std::ios_base::beg);

  return md5.hexdigest();
}